### PR TITLE
HIP and MPI+HIP Updates Part 2

### PR DESCRIPTION
**Pull request by ohearnk**
This PR backports data structure initialization fixes in PR #361 for the single GPU codepaths to the MPI+GPU codepaths.  Moreover, this PR unifies portions of the GPU and MPI+GPU initialization code to simplify future maintainence.

### CODE REVIEW

File: src/gpu/cuda/gpu.h [hunk 1]
```diff
@@ -1,112 +1,104 @@
...
 
 #include "gpu_type.h"
 
-//Madu Manathunga 07/01/2019: Libxc header files
 #include "util.h"
 #include "../xc_redistribute.h"
 
 
-// device initial and shutdown operation
-extern "C" void gpu_set_device_(int* gpu_dev_id, int* ierr);
```
Comment (agoetz) [inline src/gpu/cuda/gpu.h:20]: In general it is recommended to keep variable names in header files for code readability and documentation. Why remove them here?

**Comment by agoetz** [comment]
Note that for CUDA MPI tests with 2 tasks, DFT gradient calculations
- fail when running  on a single RTX 2080 Ti GPU
- fail or exit due to out of memory error when running on a single RTX 2080 GPU
The output files indicate an issue with XC routines:
```
 ALPHA ELECTRON DENSITY    =     0.0000000000
 BETA ELECTRON DENSITY     =     0.0000000000
```
On the RTX 2080, the program quits with an appropriate GPU out of memory error message (thanks to @ohearnk for implementing this feature!) for test `grad_psb3_b3lyp_631gss`
```
[ERROR] GPU error: cudaMemsetAsync failure
  [INFO] At line 70 in file /scratch/agoetz/QUICK/src/gpu/cuda/../gpu_get2e_getxc_drivers.h on MPI processor 1
  [INFO] Error code: 2
  [INFO] Error message: out of memory
```
This error message is unfortunately suppressed when running the test scripts (I ran the test manually to check the error message). If possible we should not suppress error messages to facilitate debugging in particular for end users.

We probably have hit an edge case where there is sufficient memory for ERIs but parts of the XC code memory handling does not use the new GPU wrappers, so the code continues and produces wrong XC results. We should open an issue for this.

**Comment by agoetz** [comment]
The full GPU test suite passes for CUDA MPI with 2 tasks when using 2 RTX 2080 Ti GPUs.
(tested on chinotto with gcc 10.2.1, openmpi 4.0.5. and cuda 12.0)

So to me this PR looks good, except I am not sure whether we really should remove variable names from header files.

File: src/gpu/cuda/gpu.h (hunk 1 above)
Comment (ohearnk) [reply src/gpu/cuda/gpu.h:20]: In most C/C++ codes which I've seen and worked on, it isn't common to use variable names in function interfaces for documentation -- other documentation mechanisms are employed, typically based on code comments (e.g., doxygen).  Moreover, for external interfaces, variable names bleed implementation details into places which should be implementation-agnostic (and thereby is also an additional places which needs to be updated if the underlying function code is modified).  This point about leaking implementation details is particularly relevent for properierty codes.

TL;DR -- it's a philosphocical software engineering choice.  Moreover, as these header files are internal interfaces in QUICK, only developers (i.e., us) will likely care.  If you feel strongly about documenting function interfaces in C/C++ code, we could also pursue using something like doxygen in place of this convention, for both internal and external interfaces.

Comment (ohearnk) [reply src/gpu/cuda/gpu.h:20]: As a data point, see the header files distributed with CUDA.

**Comment by ohearnk** [comment]
> We probably have hit an edge case where there is sufficient memory for ERIs but parts of the XC code memory handling does not use the new GPU wrappers, so the code continues and produces wrong XC results. We should open an issue for this.
> 
> QUICK should use 1 MPI task per GPU. So this is in general not a problem. We should still open an issue.

@agoetz: I ported the GPU wrappers from the OEI+ERI codes to the XC code.  Can you pull in the new commit & build to see what  that shows for the single GPU RTX 2080 TI gradient test?  If it's anything interesting, please open an issue with further details.

**Comment by agoetz** [comment]
@ohearnk I pulled in the changes with GPU wrappers in the xc code. Behavior did not change. I still see the same wrong results when runnnig two MPI tasks on a single RTX 2080 Ti card. Something else is going on.

**Comment by agoetz** [comment]
MPI on two cards and single GPU version on one card passes tests. (I did not compile/test F functions and did not run the full test suite).

File: src/gpu/cuda/gpu.h (hunk 1 above)
Comment (agoetz) [reply src/gpu/cuda/gpu.h:20]: OK. I am used to see variable names in function declarations, but I have not worked too much with C/C++. As far as I can tell this is also the case for CUDA headers (I just looked at $CUDA_HOME/include/ nvblas.h).

In any event, we don't have time now - this is a discussion for another time.
